pkg/reconciler/build/build.go

```diff
@@ -109,29 +125,151 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+				}
+
+				build.Status.SetCondition(&duckv1alpha1.Condition{
+					Type:    v1alpha1.BuildSucceeded,
+					Status:  corev1.ConditionFalse,
+					Reason:  "BuildTimeout",
+					Message: timeoutMsg,
+				})
+				// update build completed time
+				build.Status.CompletionTime = metav1.Now()
+
+				if _, err := c.updateStatus(build); err != nil {
+					c.Logger.Errorf("Failed to update status for pod: %v", err)
+					return err
+				}
 
+				c.Logger.Errorf("Timeout: %v", timeoutMsg)
+				return nil
+			}
+
+			// if not timed out then wait async
+			go c.waitForOperation(build, op)
+		} else {
+			build.Status.Builder = c.builder.Builder()
+			// If the build hasn't even started, then start it and record the operation in our status.
+			// Note that by recording our status, we will trigger a reconciliation, so the wait above
+			// will kick in.
+			var tmpl v1alpha1.BuildTemplateInterface
+			if build.Spec.Template != nil {
+				if build.Spec.Template.Kind == v1alpha1.ClusterBuildTemplateKind {
+					tmpl, err = c.clusterBuildTemplatesLister.Get(build.Spec.Template.Name)
+					if err != nil {
+						// The ClusterBuildTemplate resource may not exist.
+						if errors.IsNotFound(err) {
+							runtime.HandleError(fmt.Errorf("cluster build template %q does not exist", build.Spec.Template.Name))
+						}
+						return err
+					}
+				} else {
+					tmpl, err = c.buildTemplatesLister.BuildTemplates(namespace).Get(build.Spec.Template.Name)
+					if err != nil {
+						// The BuildTemplate resource may not exist.
+						if errors.IsNotFound(err) {
+							runtime.HandleError(fmt.Errorf("build template %q in namespace %q does not exist", build.Spec.Template.Name, namespace))
+						}
+						return err
+					}
+				}
+			}
+			build, err = builder.ApplyTemplate(build, tmpl)
+			if err != nil {
+				return err
+			}
+			// TODO: Validate build except steps+template
+			b, err := c.builder.BuildFromSpec(build)
+			if err != nil {
+				return err
+			}
+			op, err := b.Execute()
+			if err != nil {
+				build.Status.SetCondition(&duckv1alpha1.Condition{
+					Type:    v1alpha1.BuildSucceeded,
+					Status:  corev1.ConditionFalse,
+					Reason:  "BuildExecuteFailed",
+					Message: err.Error(),
+				})
+
+				if _, err := c.updateStatus(build); err != nil {
+					return err
+				}
+				return err
+			}
+			if err := op.Checkpoint(build, &build.Status); err != nil {
+				return err
+			}
+			build, err = c.updateStatus(build)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Reconciler) waitForOperation(build *v1alpha1.Build, op builder.Operation) error {
+	status, err := op.Wait()
+	if err != nil {
+		c.Logger.Errorf("Error while waiting for operation: %v", err)
+		return err
+	}
+	build.Status = *status
+	if _, err := c.updateStatus(build); err != nil {
+		c.Logger.Errorf("Error updating build status: %v", err)
+		return err
+	}
 	return nil
 }
+
+func (c *Reconciler) updateStatus(u *v1alpha1.Build) (*v1alpha1.Build, error) {
```

Comment:
At some point, you should pick up the pattern from `knative/serving` where we have a `Reconcile()` and `reconcile()` and the former just calls the latter, and then calls `updateStatus` regardless of whether the latter returned an `error`.

Reply:
Yeah, I like that pattern. I think that pattern opens up some more cleanup opportunity that I think I'd like to keep in a separate PR.